### DISCUSSION TITLE
Which cross-attention optimization technique is best? Could someone please clarify when to use it and why?

### DISCUSSION THREAD

**Discussion by superprohero** [vladmandic/sdnext#846] [Q&A]
![ugsyjuihgujksghcd](https://github.com/vladmandic/automatic/assets/133063465/509f97c4-db8f-42a0-98cf-e7a2dbc2805d)

simple question, can someone explain me please?

**Comment by vladmandic** [comment]
simple answer - it depends :)

first was split-attention, then came invokeai, then doggetx. those 3 are somewhat platform agnostic, so they can be used regardless.
and between them, i guess its in that exact order from lower to better.

xformers were heavily optimized for torch+cuda on nvidia, so if you have that, they used to be the best.
amd users are reporting that sub-quadratic works fine for them.
and sdp is newest as its built-in in torch 2.0 and typically gives similar performance to xformers with less fuss. but again, it varies depending on exact cpu and gpu (xformers does some work on cpu, so low end gpus can benefit from that while sdp as all-in on gpu, so high-end gpus work nicely).

and then second line is optional - since cross-optimization can lead to non-derministic results (tiny differences in images even if settings are the same) - this enables deterministic mode as some small cost.

if there was a simple answer, it would be pre-set and there wouldn't be a configurable option.

**Reply by superprohero** [reply]
Thanks for the simple reply, and I get what you are saying.

I have an Rtx 3060 Ti with 8 GB of RAM and a Ryzen 5600x processor.
In terms of speed, I really can't tell any difference between Xformers and SDP. The only difference is that with SDP I can't do 1440x1440 in Hires fix.

One more question: is there another way to increase the speed of the process of image creation?

**Reply by vladmandic** [reply]
increase from what? if you run benchmark, what it's do you get?

**Reply by superprohero** [reply]
Hires fix resize: from 512x512 to 1024x1024

around 20 to 22 seconds
![sfaddda](https://github.com/vladmandic/automatic/assets/133063465/1c9bba2a-74ce-4126-92dd-e8eea0180c43)

image above, can it go faster without using more ram?

**Reply by vladmandic** [reply]
hires fix is quite limited, you may want to experiment with different methods

**Comment by CapsAdmin** [comment]
On a RX 6900 XT I find that sub quadratic is the slowest, but uses the least amount of vram. This is useful when you want to generate something high resolution without getting OOM issues. In my specific case it would actually be great if it could switch between SDP and sub quadratic depending on predicted memory usage or simply just above some certain resolution.
